### PR TITLE
GH-4545: Fix String Escaping Vulnerability in FilterExpressionConverter

### PR DESCRIPTION
**Pull request by demo-zexuan**
## Summary

This PR fixes a security vulnerability (Issue [#4545](https://github.com/spring-projects/spring-ai/issues/4545)) in `AbstractFilterExpressionConverter` and its subclasses where special characters in string values were not properly escaped, leading to potential injection attacks in filter expressions.

Fixes #4545 

## Problem Description

 **Issue**: [#4545](https://github.com/spring-projects/spring-ai/issues/4545) - Many AbstractFilterExpressionConverters incorrectly handle String values

## Vulnerability Analysis

As Issue #4545 say, While investigating the codebase, I noticed that the `doSingleValue()` method in `AbstractFilterExpressionConverter` was using raw string concatenation without escaping special characters:

```java
// VULNERABLE CODE (line 149)
protected void doSingleValue(Object value, StringBuilder context) {
    if (value instanceof String) {
        context.append(String.format("\"%s\"", value));  // ⚠️ No escaping!
    }
    else {
        context.append(value);
    }
}
```

## Security Risk

This vulnerability allows attackers to:

1. **Bypass Access Control**: Inject quotes to manipulate filter logic
2. **SQL Injection**: In SQL-based vector stores (MariaDB)
3. **NoSQL Injection**: In JSON-based stores (Weaviate, MongoDB Atlas)
4. **Expression Injection**: In OData/SpEL-based stores (Azure, SimpleVectorStore)
…

## Attack Example

**Normal Request**:
```java
Expression expr = new Expression(EQ, new Key("author"), new Value("admin"));
// Output: {"author": {"$eq": "admin"}} ✅
```

**Malicious Request** (Before Fix):
```java
String malicious = "admin\" OR valueText:\"hacker";
Expression expr = new Expression(EQ, new Key("author"), new Value(malicious));
// Output: {"author": {"$eq": "admin"} OR valueText:"hacker"} ❌
// Result: FILTER BYPASSED - returns all documents!
```

**After Fix**:
```java
// Output: {"author": {"$eq": "admin\" OR valueText:\"hacker"}} ✅
// Result: Exact literal match - attack neutralized
```

What I observed seemed to be like this~

### Test Categories

1. **Basic Escaping**: Double quotes, backslashes, single quotes
2. **Combined Characters**: Mixed special characters
3. **Boundary Cases**: Empty strings, special-only strings, long strings
4. **Injection Attacks**: Quote injection, logic operator injection, nested injection
5. **Unicode**: Pure Unicode, Unicode with special characters
6. **Real-world Scenarios**: File paths, JSON snippets, SQL fragments

### Test Coverage

| Test File | Tests | Coverage |
|-----------|-------|----------|
| AbstractFilterExpressionConverterSecurityTests | 17 | Base class + JSON escaping |
| WeaviateFilterExpressionConverterSecurityTests | 6 | Weaviate-specific escaping |
| MariaDBFilterExpressionConverterSecurityTests | 6 | SQL-specific escaping |
| AzureAiSearchFilterExpressionConverterSecurityTests | 5 | OData-specific escaping |
| SimpleVectorStoreFilterExpressionConverterSecurityTests | 5 | SpEL-specific escaping |
| **Total** | **39** | **All passing** ✅ |

Thanks to LLMs, you don't need to write so many cases yourself😂

#### Source Code (5 files)

1. **AbstractFilterExpressionConverter.java**  Core fix !!!!!
   - Added `escapeStringValue()` method
   - Modified `doSingleValue()` to use escaping
   - **Impact**: Automatically protects 19 inheriting subclasses

2. **WeaviateFilterExpressionConverter.java**
   - Applied `escapeStringValue()` to both `valueText` and `valueDate`
   - **Format**: JSON-style escaping

3. **MariaDBFilterExpressionConverter.java**
   - Added `escapeSqlStringValue()` method
   - **Format**: SQL escaping (backslash + single quote doubling)

4. **AzureAiSearchFilterExpressionConverter.java**
   - Added `escapeODataStringValue()` method
   - **Format**: OData escaping (single quote doubling)

5. **SimpleVectorStoreFilterExpressionConverter.java**
   - Added `escapeSpELStringValue()` method
   - **Format**: SpEL escaping (single quote doubling)

## Escaping Rules

**THIS IS VERY IMPORTANT**

Different query languages require different escaping:

| Format | Delimiter | Escape Rules |
|--------|-----------|--------------|
| **JSON** | `"` | `\` → `\\`<br>`"` → `\"` |
| **SQL** | `'` | `\` → `\\`<br>`'` → `''` |
| **OData** | `'` | `'` → `''` |
| **SpEL** | `'` | `'` → `''` |

### Test Execution

All tests pass locally:
```bash
# Base class tests
./mvnw test -Dtest=AbstractFilterExpressionConverterSecurityTests -pl spring-ai-vector-store
# Result: PASS

# Weaviate tests
./mvnw test -Dtest=WeaviateFilterExpressionConverterSecurityTests -pl vector-stores/spring-ai-weaviate-store
# Result: PASS

# MariaDB tests
./mvnw test -Dtest=MariaDBFilterExpressionConverterSecurityTests -pl vector-stores/spring-ai-mariadb-store
# Result: PASS

# Azure tests
./mvnw test -Dtest=AzureAiSearchFilterExpressionConverterSecurityTests -pl vector-stores/spring-ai-azure-store
# Result: PASS

# SimpleVectorStore tests
./mvnw test -Dtest=SimpleVectorStoreFilterExpressionConverterSecurityTests -pl spring-ai-vector-store
# Result: PASS

### CODE REVIEW

File: vector-stores/spring-ai-mariadb-store/src/main/java/org/springframework/ai/vectorstore/mariadb/MariaDBFilterExpressionConverter.java [hunk 1]
```diff
@@ -47,14 +47,34 @@ protected void doExpression(Expression expression, StringBuilder context) {
...
+	 * unintended escape sequences.
+	 * @param input the string to escape
+	 * @return the escaped string safe for use in SQL string literals
+	 * @author Zexuan Peng &lt;pengzexuan2001@gmail.com&gt;
+	 */
+	private String escapeSqlStringValue(String input) {
+		// Replace in order: \ first, then '
+		// Backslash MUST be first to avoid double-escaping
+		// In SQL, single quote is escaped by doubling it
+		return input.replace("\\", "\\\\").replace("'", "''");
```
[the remark this answers is not in the document]
Comment (demo-zexuan) [reply vector-stores/spring-ai-mariadb-store/src/main/java/org/springframework/ai/vectorstore/mariadb/MariaDBFilterExpressionConverter.java:75]: In the original Issue #4545, the main focus was on injection attacks that might be introduced by quotes, while control character escaping is an additional security enhancement and not the core objective of this fix. Copilot's suggestion goes beyond the scope of the original issue, so ~waiting for administrator.

File: spring-ai-vector-store/src/main/java/org/springframework/ai/vectorstore/filter/converter/AbstractFilterExpressionConverter.java [hunk 2]
```diff
@@ -208,6 +210,35 @@ protected void doAddValueRangeSpitter(Filter.Value listValue, StringBuilder cont
...
+	 * their vector store implementation.
+	 *
+	 * <p>
+	 * <b>Note:</b> This method escapes backslashes and double quotes as required by
+	 * JSON-like filter syntax. Single quotes are NOT escaped as they are not special
+	 * characters in JSON. Subclasses targeting vector stores with different escaping
+	 * requirements (e.g., SQL-based stores) should override this method.
+	 * @param input the string to escape
+	 * @return the escaped string safe for use in filter expressions
+	 * @author Zexuan Peng <pengzexuan@gmail.com>
```
[the remark this answers is not in the document]
Comment (demo-zexuan) [reply spring-ai-vector-store/src/main/java/org/springframework/ai/vectorstore/filter/converter/AbstractFilterExpressionConverter.java]: In the original Issue #4545, the main focus was on injection attacks that might be introduced by quotes, while control character escaping is an additional security enhancement and not the core objective of this fix. Copilot's suggestion goes beyond the scope of the original issue, so ~waiting for administrator.

**Comment by demo-zexuan** [comment]
In the original Issue #4545, the main focus was on injection attacks that might be introduced by quotes, while control character escaping is an additional security enhancement and not the core objective of this fix. Copilot's suggestion goes beyond the scope of the original issue, so ~waiting for administrator.

**Comment by ilayaperumalg** [comment]
@demo-zexuan Closing this PR as we already are tracking this internally. Patch for this will be available soon.

**Comment by demo-zexuan** [comment]
> @demo-zexuan Closing this PR as we already are tracking this internally. Patch for this will be available soon.

Okay, thank you for your hard work.
